File: Python/BackTracking/16_longest_possible_route.py
"""
Idea:
Using Recursion & backtracking, with each cell in the matrix. if end is reached, saving it-till all paths are discovered
returning the longest path's length
"""


class Solution:
    def __init__(self):
        # Initialize class variables
        self.row = 0
        self.col = 0
        self.ans = -1

    def solve(self, mat, target_x, target_y, x, y, dist):
        # Base cases for recursion
        if x < 0 or y < 0 or x == self.row or y == self.col or mat[x][y] != 1:
            return
        # Reached the target cell, update the answer
        if x == target_x and y == target_y:
            self.ans = max(self.ans, dist)
            return

        # Mark the current cell as visited
        mat[x][y] = -1

        # Explore all four possible directions
        self.solve(mat, target_x, target_y, x + 1, y, dist + 1)
        self.solve(mat, target_x, target_y, x - 1, y, dist + 1)
        self.solve(mat, target_x, target_y, x, y + 1, dist + 1)
        self.solve(mat, target_x, target_y, x, y - 1, dist + 1)

        # Backtrack: restore the original value of the current cell
        mat[x][y] = 1

    def longestPath(self, mat, n: int, m: int, xs: int, ys: int, xd: int, yd: int) -> int:
        # Set the dimensions of the matrix
        self.row = n
        self.col = m

        # Start the recursive exploration from the source cell
        self.solve(mat, xd, yd, xs, ys, 0)

        # Return the final result
        return self.ans


# {
# Driver Code Starts

class IntArray:

    def __init__(self) -> None:
        pass

    def Input(self, n):
        arr = [int(i) for i in input().strip().split()]  # array input
        return arr

    def Print(self, arr):
        for i in arr:
            print(i, end=" ")
        print()


class IntMatrix:
    def __init__(self) -> None:
        pass

    def Input(self, n, m):
        matrix = []
        # matrix input
        for _ in range(n):
            matrix.append([int(i) for i in input().strip().split()])
        return matrix

    def Print(self, arr):
        for i in arr:
            for j in i:
                print(j, end=" ")
            print()


if __name__ == "__main__":
    t = int(input())
    for _ in range(t):
        a = IntArray().Input(2)

        b = IntArray().Input(4)

        mat = IntMatrix().Input(a[0], a[0])

        obj = Solution()
        res = obj.longestPath(mat, a[0], a[1], b[0], b[1], b[2], b[3])

        print(res)

# } Driver Code Ends

File: C++/BackTracking/16_longest_possible_route.cpp
/*
Idea:
Using Recursion & backtracking, with each cell in the matrix. if end is reached, saving it-till all paths are discovered
returning the longest path's length
*/

//{ Driver Code Starts
//Initial Template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
//User function Template for C++

class Solution {
    int row, col, ans = -1;

public:
    // Helper function to recursively explore the matrix and find the longest path
    void solve(vector<vector<int>>& mat, int target_x, int target_y, int x, int y, int dist)
    {
        // Base cases for recursion
        if (x < 0 || y < 0 || x == row || y == col || mat[x][y] != 1)
            return;

        // Reached the target cell, update the answer
        if (x == target_x && y == target_y)
        {
            ans = max(ans, dist);
            return;
        }

        // Mark the current cell as visited
        mat[x][y] = -1;

        // Explore all four possible directions
        solve(mat, target_x, target_y, x + 1, y, dist + 1);
        solve(mat, target_x, target_y, x - 1, y, dist + 1);
        solve(mat, target_x, target_y, x, y + 1, dist + 1);
        solve(mat, target_x, target_y, x, y - 1, dist + 1);

        // Backtrack: restore the original value of the current cell
        mat[x][y] = 1;
    }

    // Main function to find the longest path in the matrix
    int longestPath(vector<vector<int>>& matrix, int xs, int ys, int xd, int yd)
    {
        // Set the dimensions of the matrix
        row = matrix.size();
        col = matrix[0].size();

        // Start the recursive exploration from the source cell
        solve(matrix, xd, yd, xs, ys, 0);

        // Return the final result
        return ans;
    }
};

//{ Driver Code Starts.

int main()
{
    int t;
    cin >> t;
    while (t--)
    {
        int n, m;
        cin >> n >> m;
        int xs, ys, xd, yd;
        cin >> xs >> ys >> xd >> yd;
        vector<vector<int>> mat(n, vector<int>(m));
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                cin >> mat[i][j];
            }
        }
        Solution ob;
        int ans = ob.longestPath(mat, xs, ys, xd, yd);
        cout << ans;
        cout << "\n";
    }
    return 0;
}
// } Driver Code Ends